File: FreeTypeStrategy.cpp
#include <string>
#include <iostream>
#include <cstdint>
#include "math.h"
#include "FreeTypeStrategy.h"

// EMSCRIPTEN WASM
#include <emscripten.h>

// FREETYPE
#include <ft2build.h>
#include FT_FREETYPE_H

FT_Library library;
FT_Face face;
FT_Error error;
FT_GlyphSlot glyph;

//const std::string FONT_PATH = "fonts/";
//const std::string TTF_FILE_EXTENSION = ".ttf";

int InitializeFreeTypeC() {
    // Initialize FreeType
    error = FT_Init_FreeType(&library);
    if (error) {
        std::cerr << "Error initializing FreeType" << std::endl;
    }

    return error;
}

int LoadFontC(char* font) //$quorum_text = function(fontName)
{
    // Make sure requested font is arial
    //if (font != "Arial") {
    char arialFont[] = "Arial";
    if (strcmp(font, arialFont) != 0) {
        std::cerr << "Only the arial font face is supported at this time." << std::endl;
        return 1;
    }
    else {
        std::cerr << "'Arial' passed to LoadFontNative." << std::endl;
    }

    // Load font
    error = FT_New_Face(library, "fonts/Arial.ttf", 0, &face);
    if (error == FT_Err_Unknown_File_Format) {
        std::cerr << "Font format is unsupported" << std::endl;
        return 1;
    }
    else if (error) {
        std::cerr << "Font file is missing or corrupted" << std::endl;
        return 1;
    }

    //SetSizeC(36); // temporary, but needed to render glyph
    return 0;
}

int IsFontAvailableC(char* font) //$quorum_text = function(fontName)
{
    char arialFont[] = "Arial"; // Only using Arial right now...
    return strcmp(font, arialFont); // Return 1 (True) if font == 'Arial'
}

void SetSizeC(int size) //$quorum_integer = function(size)
{
    error = FT_Set_Char_Size(face, size << 6, 0, 72, 0);
}

void SetAngleC(double angle) //$quorum_number = function(angle)
{
    std::cerr << "Angle: " << angle << std::endl;
    FT_Matrix matrix;

    matrix.xx = (FT_Fixed)(cos(angle) * 0x10000L);
    matrix.xy = (FT_Fixed)(-sin(angle) * 0x10000L);
    matrix.yx = (FT_Fixed)(sin(angle) * 0x10000L);
    matrix.yy = (FT_Fixed)(cos(angle) * 0x10000L);

    FT_Set_Transform(face, &matrix, 0);
}


void DisposeC() // = function()
{
    FT_Done_Face(face);
}

int GetKerningC(char* current, char* next) //$quorum_text$quorum_text = function(currentCharacter, nextCharacter)
{
    if (FT_HAS_KERNING(face)) {
        FT_Vector delta;
        FT_UInt current_index = FT_Get_Char_Index(face, current[0]);
        FT_UInt next_index = FT_Get_Char_Index(face, next[0]);
        FT_Get_Kerning(face, current_index, next_index,
            FT_KERNING_DEFAULT, &delta);
        return (int)(delta.x >> 6);
    } else
        return 0;
}

long GetUnderlinePositionC() // = function()
{
    return FT_MulFix(face->underline_position, face->size->metrics.y_scale);
}

long GetUnderlineThicknessC() // = function()
{
    return FT_MulFix(face->underline_thickness, face->size->metrics.y_scale);
}

int GetLineHeightC() // = function()
{
    return face->size->metrics.height;
}

void GetAvailableFontsC() // = function()
{
    // NYI
    //return null;
}

int GetMaximumAscentC() // = function()
{
    return face->size->metrics.ascender;
}

int GetMaximumDescentC() // = function()
{
    return face->size->metrics.descender;
}

int loadCharC(char* symbol)
{
    char sym = *symbol;
    std::cerr << "Loading glyph:" << sym << std::endl;
    glyph = face->glyph;

    // If an error occurs during FT_Load_Char, return 1.
    error = FT_Load_Char(face, sym, FT_LOAD_RENDER);
    if (error) {
        std::cerr << "Error: FT_Load_Char " << error << std::endl;
        return 1;
    }
    else {
        return 0;
    }
}

void getBitmapC(unsigned int output_ptr, int num_bytes) {
    // Get FreeType bitmap buffer
    unsigned char* buf = glyph->bitmap.buffer;
    // Set output pointer
    uint8_t* dst = (uint8_t*)output_ptr;
    // Write data to output
    for (int i = 0; i < num_bytes; i++) {
        dst[i] = (uint8_t)buf[i];
        std::cerr << (int)buf[i] << " ";
    }
    std::cerr << std::endl;
}

int getBitmapDataC(long* bitmapData)
{
    bitmapData[0] = glyph->bitmap_left;
    bitmapData[1] = glyph->bitmap_top;
    bitmapData[2] = glyph->bitmap.rows;
    bitmapData[3] = glyph->bitmap.width;
    bitmapData[4] = glyph->advance.x;
    bitmapData[5] = glyph->advance.y;
    bitmapData[6] = glyph->bitmap.pitch;

    std::cerr << "WASM - Left: " << bitmapData[0] << " Top: " << bitmapData[1] << " Rows: " << bitmapData[2] << " Width: " << bitmapData[3] << " X: " << bitmapData[4] << " Y: " << bitmapData[5] << " Pitch: " << bitmapData[6] << std::endl;
    
    return 0;
}

int main() {
    return InitializeFreeTypeC();
}

File: FreeTypeStrategy.js
function plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy_(quorumFont) {
	if (!plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy_.initialized_plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy_) {
		// Static initializer for the class.
		console.log('Static initializer for plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy');

		/*****************************************************
		 * Run main() which initializes FreeType
		 ****************************************************/

		main_return = Module.cwrap('main', 'number')
		console.log("execute: main() with return value = " + main_return());

		// Plug-In loaded = true
		plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy_.initialized_plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy_ = true;
	}

	var color = null;
	this.me_ = quorumFont;

	this.LoadFontNative$quorum_Libraries_System_File = function (file) {
		// NYI
		console.log('LoadFontNative$quorum_Libraries_System_File = function(file)');
	};

	this.LoadFontNative$quorum_text = function (fontName) {
		// NYI
		console.log('LoadFontNative$quorum_text = function(fontName)');
		console.log('fontName = ' + fontName);

		var loadFontNativeC = Module.cwrap('LoadFontC', 'number', ['string']);
		var loadFontNativeResult = loadFontNativeC(fontName);
		if (loadFontNativeResult == 1) {
			// error loading font
			console.log('Error loading font.');
		} else {
			console.log('Font loaded succesfully.');
		}
		console.log("Font size: " + this.me_.GetSize());
		console.log("Font angle: " + this.me_.GetAngle());
		this.SetSizeNative$quorum_integer(this.me_.GetSize());
		if (this.me_.GetAngle() != 0)
			this.SetAngleNative$quorum_number(this.me_.GetAngle());
	};

	this.IsFontAvailable$quorum_text = function (fontName) {
		// NYI
		console.log('IsFontAvailable$quorum_text = function(fontName)');
		return true;
	};

	this.SetSizeNative$quorum_integer = function (size) {
		console.log('SetSizeNative$quorum_integer = function(size)');
		var SetSizeC = Module.cwrap('SetSizeC', null, ['number']);
		SetSizeC(size);
	};

	this.SetAngleNative$quorum_number = function (angle) {
		console.log('SetAngleNative$quorum_number = function(' + angle + ')');
		var SetAngleC = Module.cwrap('SetAngleC', null, ['number']);
		SetAngleC(angle / 180 * Math.PI);
		//SetAngleC(angle);

	};

	this.SetColorNative$quorum_Libraries_Game_Graphics_Color = function (newColor) {
		color = newColor;
		console.log('SetColorNative$quorum_Libraries_Game_Graphics_Color = function(newColor)');
	};

	this.GetColor = function () {
		console.log('GetColor = function()');
		return color;
	};

	this.DisposeNative = function () {
		console.log('DisposeNative = function()');
		var DisposeC = Module.cwrap('DisposeC', null, null);
		DisposeC();
	};

	this.GetGlyphNative$quorum_text = function (character) {
		console.log('GetGlyphNative$quorum_text = function(' + character + ')');
		// Create glyph
		var glyph = new quorum_Libraries_Game_Graphics_Glyph_();

		// Load character into FreeType
		var loadChar = Module.cwrap('loadCharC', 'number', ['string']);
		var loadCharResult = loadChar(character);
		if (loadCharResult == 1) {
			// error loading glyph
			console.log('Error loading glyph.');
		} else {
			console.log('Glyph loaded succesfully.');
		}

		// Create data array for glyph metrics (7 * 4 bytes = 28 bytes)
		var offset = Module._malloc(28);
		console.log(offset);

		// Setup call to getBitmapData from WASM module
		var getBitmapData = Module.cwrap('getBitmapDataC', 'number', ['number']);

		// Pass array pointer to getBitmapData
		getBitmapData(offset);

		// Create array and copy glyph metrics into it
		var bitmapData = [];
		bitmapData[0] = Module.getValue(offset, 'i32'); //glyph->bitmap_left;
		bitmapData[1] = Module.getValue(offset + 4, 'i32'); //glyph->bitmap_top;
		bitmapData[2] = Module.getValue(offset + 8, 'i32'); //glyph->bitmap.rows;
		bitmapData[3] = Module.getValue(offset + 12, 'i32'); //glyph->bitmap.width;
		bitmapData[4] = Module.getValue(offset + 16, 'i32'); //glyph->advance.x;
		bitmapData[5] = Module.getValue(offset + 20, 'i32'); //glyph->advance.y;
		bitmapData[6] = Module.getValue(offset + 24, 'i32'); //glyph->bitmap.pitch;

		console.log(" JS  - Left: " + bitmapData[0] + " Top: " + bitmapData[1] + " Rows: " + bitmapData[2] + " Width: " + bitmapData[3] + " X: " + bitmapData[4] + " Y: " + bitmapData[5] + " Pitch: " + bitmapData[6]);

		// Free allocated memory
		Module._free(offset);

		// Calculate size of byte array for bitmap pixel buffer
		// size = |pitch|*rows
		var size = bitmapData[6];
		if (size < 0)
			size = -size;
		size *= bitmapData[2];

		// Allocate memory on Emscripten heap
		let num_bytes = size;
		let dataPtr = Module._malloc(num_bytes);
		let dataHeap = new Uint8Array(Module.HEAPU8.buffer, dataPtr, num_bytes);
		console.log('Allocated memory on heap.');

		// Call WASM function
		console.log('Calling WASM function "getBitmap"...');
		var getBitmap = Module.cwrap('getBitmapC', null, ['number', 'number']);
		getBitmap(dataHeap.byteOffset, num_bytes);

		// Get data from Emscripten heap
		let bitmapBuffer = new Uint8Array(dataHeap.buffer, dataHeap.byteOffset, num_bytes);

		// Output buffer to log
		var i;
		var output = "";
		for (i = 0; i < size; i++) {
			output = output + bitmapBuffer[i].toString(10) + " ";
		}
		//console.log('out_data: ' + bitmapBuffer[0] + ', ' + bitmapBuffer[1] + ', ..., ' + bitmapBuffer[size-1]);
		console.log(output);
		console.log(bitmapBuffer);

		// Free allocated memory
		Module._free(dataHeap.byteOffset);

		if (character != ' ') {
			// Create Alpha pixel map
			var pixmap = new quorum_Libraries_Game_Graphics_PixelMap_();
			var map = pixmap.plugin_;
			var format = new quorum_Libraries_Game_Graphics_Format_();
			format.SetValue$quorum_integer(format.Get_Libraries_Game_Graphics_Format__ALPHA_());
			map.LoadFromFontBitmap(bitmapBuffer, bitmapData[3], bitmapData[2], format);


			console.log("'" + character + "' Pixelmap Width: " + pixmap.GetWidth() + " Expected: " + bitmapData[3]);
			console.log("'" + character + "' Pixelmap Height: " + pixmap.GetHeight() + " Expected: " + bitmapData[2]);
			console.log(map.GetPixels());

			// Create File texture data
			var texData = new quorum_Libraries_Game_Graphics_FileTextureData_();
			texData.InitializeFileTextureData$quorum_Libraries_System_File$quorum_Libraries_Game_Graphics_PixelMap$quorum_Libraries_Game_Graphics_Format$quorum_boolean(null, pixmap, null, false);
			//texData.SetDisposalState(false);

			// Create texture
			var texture = new quorum_Libraries_Game_Graphics_Texture_();
			texture.LoadFromTextureData$quorum_Libraries_Game_Graphics_TextureData(texData);

			// Create color
			var c = new quorum_Libraries_Game_Graphics_Color_();
			c.SetColor$quorum_number$quorum_number$quorum_number$quorum_number(color.GetRed(), color.GetGreen(), color.GetBlue(), color.GetAlpha());
			texture.plugin_.fontColor = c;

			// Create texture region
			var region = new quorum_Libraries_Game_Graphics_TextureRegion_();
			region.LoadTextureRegion$quorum_Libraries_Game_Graphics_Texture(texture);

			glyph.texture = region;
		} else {
			glyph.texture = null;
		}

		//glyph.Set_Libraries_Game_Graphics_Glyph__texture_(plugins_quorum_Libraries_Game_Graphics_Fonts_FreeTypeStrategy_.testDrawable);
		glyph.Set_Libraries_Game_Graphics_Glyph__horizontalAdvance_(bitmapData[4] >> 6);
		glyph.Set_Libraries_Game_Graphics_Glyph__verticalAdvance_(bitmapData[5] >> 6);
		glyph.Set_Libraries_Game_Graphics_Glyph__lengthToGlyph_(bitmapData[0]);
		glyph.Set_Libraries_Game_Graphics_Glyph__heightFromBaseLine_(bitmapData[1]);


		return glyph;
	};

	this.GetKerning$quorum_text$quorum_text = function (currentCharacter, nextCharacter) {
		var GetKerningC = Module.cwrap('GetKerningC', 'number', ['string', 'string']);
		return GetKerningC(currentCharacter, nextCharacter);
	};

	this.FinishedLoading = function () {
		return true;
	};

	this.LoadImageSheet$Libraries_Game_Graphics_Fonts_FontImageSheet = function (imageSheet) {
		// NYI
		console.log('LoadImageSheet$Libraries_Game_Graphics_Fonts_FontImageSheet = function(imageSheet)');
	};

	this.LoadImageSheet$quorum_Libraries_Game_Graphics_Fonts_FontImageSheet = function (imageSheet) {
		// NYI
		console.log('LoadImageSheet$quorum_Libraries_Game_Graphics_Fonts_FontImageSheet = function(imageSheet)');
		return false;
	};

	this.SystemGetHeight = function () {
		return this.GetLineHeight();
	};

	this.SystemGetMaximumAscent = function () {
		return this.GetMaximumAscent();
	};

	this.SystemGetMaximumDescent = function () {
		return this.GetMaximumDescent();
	};

	this.SystemGetUnderlinePosition = function () {
		var GetMaximumAscentC = Module.cwrap('GetMaximumAscentC', 'number', null);
		return (GetMaximumAscentC() >> 6);
	};

	this.SystemGetUnderlineThickness = function () {
		var GetUnderlinePositionC = Module.cwrap('GetUnderlinePositionC', 'number', null);
		return (GetUnderlinePositionC() >> 6);
	};

	this.GetLineHeight = function () {
		var GetUnderlineThicknessC = Module.cwrap('GetUnderlineThicknessC', 'number', null);
		return (GetUnderlineThicknessC() >> 6);
	};

	this.GetAvailableFonts = function () {
		// NYI
		console.log('GetAvailableFonts = function()');
		return null;
	};

	this.GetMaximumAscent = function () {
		var GetMaximumAscentC = Module.cwrap('GetMaximumAscentC', 'number', null);
		return (GetMaximumAscentC() >> 6);
	};

	this.GetMaximumDescent = function () {
		var GetMaximumDescentC = Module.cwrap('GetMaximumDescentC', 'number', null);
		return (GetMaximumDescentC() >> 6);
	};

	this.GetLineGap = function () {
		// NYI
		console.log('GetLineGap = function()');
	};
}